httpbakery/discharge.go

```diff
@@ -56,66 +84,122 @@ type dischargeHandler struct {
+// discharger using the key and locator from the given service.
+func NewDischargerFromService(svc *bakery.Service, checker ThirdPartyCaveatChecker) *Discharger {
+	return NewDischarger(DischargerParams{
+		Checker: checker,
+		Key:     svc.Key(),
+		Locator: svc.Locator(),
+	})
 }
 
-func (d *dischargeHandler) serveDischarge(p httprequest.Params) (interface{}, error) {
-	r, err := d.serveDischarge1(p)
-	if err != nil {
-		logger.Debugf("serveDischarge -> error %#v", err)
-	} else {
-		logger.Debugf("serveDischarge -> %#v", r)
+// NewDischarger returns a new third-party caveat discharger
+// using the given parameters.
+func NewDischarger(p DischargerParams) *Discharger {
+	if p.ErrorToResponse == nil {
+		p.ErrorToResponse = ErrorToResponse
+	}
+	if p.Locator == nil {
+		p.Locator = bakery.PublicKeyLocatorMap(nil)
+	}
+	return &Discharger{
+		p: p,
 	}
-	return r, err
 }
 
-func (d *dischargeHandler) serveDischarge1(p httprequest.Params) (interface{}, error) {
-	logger.Debugf("dischargeHandler.serveDischarge {")
-	defer logger.Debugf("}")
-	if p.Request.Method != "POST" {
-		// TODO http.StatusMethodNotAllowed)
-		return nil, badRequestErrorf("method not allowed")
-	}
-	p.Request.ParseForm()
-	id := p.Request.Form.Get("id")
-	if id == "" {
-		return nil, badRequestErrorf("id attribute is empty")
+// AddMuxHandlers adds handlers to the given ServeMux to provide
+// a third-party caveat discharge service.
+func (d *Discharger) AddMuxHandlers(mux *http.ServeMux, rootPath string) {
+	for _, h := range d.Handlers() {
+		// Note: this only works because we don't have any wildcard
+		// patterns in the discharger paths.
+		mux.Handle(path.Join(rootPath, h.Path), mkHTTPHandler(h.Handle))
 	}
-	// TODO support macaroon id independent of caveat id
+}
 
-	checker := func(info *bakery.ThirdPartyCaveatInfo) ([]checkers.Caveat, error) {
-		return d.checker(p.Request, info)
+// Handlers returns a slice of handlers that can handle a third-party
+// caveat discharge service when added to an httprouter.Router.
+func (d *Discharger) Handlers() []httprequest.Handler {
+	f := func(p httprequest.Params) (dischargeHandler, error) {
+		return dischargeHandler{
+			discharger: d,
+		}, nil
 	}
+	return httprequest.ErrorMapper(d.p.ErrorToResponse).Handlers(f)
+}
 
-	var resp dischargeResponse
-	m, err := d.svc.Discharge(bakery.ThirdPartyCheckerFunc(checker), []byte(id))
+// dischargeHandler is the type used to define the httprequest handler
+// methods for a discharger.
+type dischargeHandler struct {
+	discharger *Discharger
+}
+
+// dischargeRequest is a request to create a macaroon that discharges the
+// supplied third-party caveat. Discharging caveats will normally be
+// handled by the bakery it would be unusual to use this type directly in
+// client software.
+type dischargeRequest struct {
+	httprequest.Route `httprequest:"POST /discharge"`
+	Id                string `httprequest:"id,form"`
+	//	Id64		string `httprequest:"id64,form"`
```

Comment:
Could I leave it and save the conflicts please? It gets uncommented in the very next PR...